Sources/SwiftDocC/Infrastructure/Link Resolution/SnippetResolver.swift

```diff
@@ -0,0 +1,156 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2025 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import SymbolKit
+import Markdown
+
+/// A type that resolves snippet paths.
+final class SnippetResolver {
+    typealias SnippetMixin = SymbolKit.SymbolGraph.Symbol.Snippet
+    typealias Explanation  = Markdown.Document
+    
+    /// Information about a resolved snippet
+    struct ResolvedSnippet {
+        fileprivate var path: String // For use in diagnostics
+        var mixin: SnippetMixin
+        var explanation: Explanation?
+    }
+    /// A snippet that has been resolved, either successfully or not.
+    enum SnippetResolutionResult {
+        case success(ResolvedSnippet)
+        case failure(TopicReferenceResolutionErrorInfo)
+    }
+    
+    private var snippets: [String: ResolvedSnippet] = [:]
+    
+    init(symbolGraphLoader: SymbolGraphLoader) {
+        var snippets: [String: ResolvedSnippet] = [:]
+        
+        for graph in symbolGraphLoader.snippetSymbolGraphs.values {
+            for symbol in graph.symbols.values {
+                guard let snippetMixin = symbol[mixin: SnippetMixin.self] else { continue }
+                
+                let path: String = if symbol.pathComponents.first == "Snippets" {
+                    symbol.pathComponents.dropFirst().joined(separator: "/")
+                } else {
+                    symbol.pathComponents.joined(separator: "/")
+                }
+                
+                snippets[path] = .init(path: path, mixin: snippetMixin, explanation: symbol.docComment.map {
+                    Document(parsing: $0.lines.map(\.text).joined(separator: "\n"), options: .parseBlockDirectives)
+                })
+            }
+        }
+        
+        self.snippets = snippets
+    }
+ 
+    func resolveSnippet(path authoredPath: String) -> SnippetResolutionResult {
```

Comment:
FYI: there are some nice "Good first issue" improvements to be had here. For example, if the snippet file is nested in a deeper subdirectory but there's only one snippet with that base name, we could have a custom solution an/or note that says that a snippet with that name exist in a deeper path. After this is merged I can open an issue suggesting that.